src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -201,5 +138,203 @@ public SagaVersion(long version)
+                    }
+
+                    var correlationLockKey = CreateCorrelationLockKey(sagaType, correlationProperty.Value);
+
+                    if (!lockers.ContainsKey(correlationLockKey))
+                    {
+                        lockers.Add(correlationLockKey, lockToken);
+                    }
+                    if (!lockers.ContainsKey(sagaType.FullName))
+                    {
+                        lockers.Add(sagaType.FullName, new object());
+                    }
+
+                    sagaIdIndex.Add(sagaData.Id, new VersionedSagaEntity(sagaData));
+
+                    if (!correlationValueIndex.ContainsKey(sagaType))
+                    {
+                        correlationValueIndex.Add(sagaType, new Dictionary<string, Guid>());
+                    }
+
+                    if (correlationValueIndex[sagaType].ContainsKey(correlationProperty.Value.ToString()))
+                    {
+                        correlationValueIndex[sagaType].Remove(correlationProperty.Value.ToString());
+                    }
+
+                    correlationValueIndex[sagaType].Add(correlationProperty.Value.ToString(), sagaData.Id);
+                }
+            }
+
+            object GetLocker(Guid? sagaId, Type sagaType, object correlationValue)
+            {
+                var correlationKey = correlationValue == null ? null : CreateCorrelationLockKey(sagaType, correlationValue);
+
+                if (sagaId !=null && lockers.ContainsKey(sagaId.ToString()))
+                {
+                    return lockers[sagaId.ToString()];
+                }
+                if (correlationKey != null && lockers.ContainsKey(correlationKey))
+                {
+                    return lockers[correlationKey];
+                }
+                if (lockers.ContainsKey(sagaType.FullName))
+                {
+                    return lockers[sagaType.FullName];
+                }
+
+                return new object();
+            }
+
+            public void Update(IContainSagaData sagaData)
+            {
+                var sagaType = sagaData.GetType();
+                var correlationProperty = sagaType.GetProperty(correlationProperties[sagaType]);
+
+                var lockToken = GetLocker(sagaData.Id, sagaData.GetType(), null);
+
+                lock (lockToken)
+                {
+                    var oldSaga = sagaIdIndex[sagaData.Id];
+                    var oldCorrelationValue = correlationProperty.GetValue(oldSaga.SagaData);
+                    var newCorrelationValue = correlationProperty.GetValue(sagaData);
+
+                    sagaIdIndex[sagaData.Id] = new VersionedSagaEntity(sagaData, oldSaga);
+
+                    correlationValueIndex[sagaType].Remove(oldCorrelationValue.ToString());
+                    correlationValueIndex[sagaType].Add(newCorrelationValue.ToString(), sagaData.Id);
+                    lockers.Remove(CreateCorrelationLockKey(sagaType, oldCorrelationValue));
```

Comment:
As `lockers` aren't thread safe and `lockToken` is acquired for a specific saga, multiple threads might access this structure concurrently.